io-engine/src/subsys/nvmf/subsystem.rs

```diff
@@ -165,6 +196,189 @@ fn make_sn<T: AsRef<[u8]>>(uuid: T) -> CString {
+                debug!(
+                    "device name  {:?} not a valid NVMe controller",
+                    dev_name
+                );
+                Self::reset_cb(false, ctx);
+                return;
+            }
+        };
+        debug!("nvmf_event_handler: nvme controller {:?} ", ctrlr);
+        let mut ctrlr = ctrlr.lock();
+        if ctrlr.reset(NvmfSubsystem::reset_cb, ctx, false).is_ok() {
+            debug!("nvmf_event_handler: Reset initiated for {:?} ", ctrlr);
+        } else {
+            error!(
+                "nvmf_event_handler: Failed to initiate reset for {:?} ",
+                ctrlr
+            );
+            Self::reset_cb(false, ctx);
+        }
+    }
+
+    pub async fn reset_nexus(nexus_name: String) {
+        let mut nexus = match nexus_lookup_mut(&nexus_name) {
+            Some(v) => v,
+            None => {
+                debug!("nqn {:?} is not having valid nexus", nexus_name);
+                return;
+            }
+        };
+
+        //Pause the Nexus
+        if nexus.as_mut().pause().await.is_err() {
+            error!("Failed to pause Nexus {:?}", nexus);
+            return;
+        }
+        info!("Paused the nexus - {:?} ", nexus_name);
+        for child in nexus.children_iter() {
+            debug!("Found a {:?} for nexus {:?}", child, nexus_name);
+            if !child.is_healthy() {
+                continue;
+            }
+            let dev = match child.get_device() {
+                Ok(dev) => dev,
+                Err(error) => {
+                    debug!(
+                        "Nexus {:?} get_device failed for {:?} with Error {:?}",
+                        nexus_name, child, error
+                    );
+                    continue;
+                }
+            };
+            let dev_name = dev.device_name();
+            if dev.driver_name() != "nvme" {
+                debug!(
+                    "Device {:?} is not NVMe-oF, reset is not needed",
+                    dev_name
+                );
+                continue;
+            }
+            debug!(
+                "nvmf_event_handler: device name {:?} - nexus nexus_name {:?},
+                 driver {:?}, product {:?}",
+                dev_name,
+                nexus_name,
+                dev.driver_name(),
+                dev.product_name(),
+            );
+            let (sender, receiver) = oneshot::channel::<bool>();
+            Reactors::master().send_future(Self::reset_one_child(
+                dev_name.clone(),
+                cb_arg(sender),
+            ));
+            if !receiver.await.expect("Reset completed") {
+                error!(
+                    "Failed to reset {:?} for nexus_name {:?}",
+                    child, nexus_name
+                );
+            }
+        }
+
+        //Resume the Nexus
+        if nexus.as_mut().resume().await.is_err() {
+            error!("Failed to resume Nexus {:?}", nexus);
+            return;
```

Comment:
if we return here, we don't restore nexus state. Shall we move resume logic after code that restores the state ?

Reply:
In fact, we should not restore the state to open if the resume failed. At this point, it will need user intervention anyway. The user needs to unshare and share the nexus again

Reply:
When can resume fail? Would it be worth retrying?

Reply:
It can fail on nomem, the invalid state being passed, or busy. We have retry logic in case of a busy inside resume though It will unlikely be busy since it is already paused. Hence, it may not be useful to retry here.